### PR TITLE
Add timeout-minutes prop for building an AL-Go project

### PR DESCRIPTION
**Pull request by mazhelez**
Define a timeout for building an AL-Go project.

Default value is 180 (minutes).

### CODE REVIEW

File: Templates/AppSource App/.github/workflows/CICD.yaml [hunk 1]
```diff
@@ -207,6 +208,7 @@ jobs:
 
       - name: Read secrets
         uses: microsoft/AL-Go-Actions/ReadSecrets@main
+        timeout-minutes: ${{ fromJson(steps.readProjectSettings.outputs.SettingsJson).buildTimeoutInMinutes }}
```
[the remark this answers is not in the document]
Comment (mazhelez) [reply Templates/AppSource App/.github/workflows/CICD.yaml:211]: Oh, yes. Nice catch!

File: Actions/AL-Go-Helper.ps1 [hunk 2]
```diff
@@ -483,6 +483,7 @@ function ReadSettings {
         "doNotRunBcptTests"                      = $false
         "doNotPublishApps"                       = $false
         "doNotSignApps"                          = $false
+        "buildTimeoutInMinutes"                  = "60"
```
[the remark this answers is not in the document]
Comment (mazhelez) [reply Actions/AL-Go-Helper.ps1:486]: Yes.

If someone is running build for more than an hour, they'll need to set `buildTimeoutInMinutes` to something more appropriate.
My approach was to add a value that would work for _most_. Hence, the magic number.

[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:486]: I kind of agree that 60 minutes is a better default than 360 minutes, but I also don't want to bother with partners running into this problem (as I know somebody will)
They will create issues and I will have to tell them all that they need to create this setting.
As I see it we can do one out of two things:
1.  Set the default to 360 and then add a few lines of code to Run-Pipeline action, which measures the time used, and if the time is larger than 60 minutes, then add a warning explaining that they should add this setting, as 60 minutes will be the default in a one of the upcoming releases.
2. Add a step, which runs if the runpipeline times out (not sure that is doable) and explain what they need to change.

[the remark this answers is not in the document]
Comment (mazhelez) [reply Actions/AL-Go-Helper.ps1:486]: Something something like the following right after `RunPipeline`:

```
      - name: Timeout Warning
        if: failure() && steps.RunPipeline.conclusion = 'timed_out'
        run: |
          $currentProject = ${{ matrix.project }}
          $currentTimeout = ${{ fromJson(steps.readProjectSettings.outputs.SettingsJson).buildTimeoutInMinutes }}

          if($currentTimeout -eq 60) { #if the current timeout is the default (60), the assumption is that it was not customized.
            Write-Host "::Warning::Current timeout for project $currentProject is set to $currentTimeout (minutes). Consider increasing it by specifying AL-Go setting buildTimeoutInMinutes."
          }
```

But now the question is... when would we remove such warning?

[the remark this answers is not in the document]
Comment (aholstrup1) [reply Actions/AL-Go-Helper.ps1:486]: I think we should just print a timeout warning regardless of whether the current timeout is the default one.

[the remark this answers is not in the document]
Comment (mazhelez) [reply Actions/AL-Go-Helper.ps1:486]: Increasing the timeout might not be the proper solution is some cases. Timeouts could potentially reveal issues with locks and hanging processes. 

In our case we want to show a warning, as we change the default (from 360 to 60 minutes). Which isn't even the case as the 360 minutes is for the whole job.

Another solution would be to document out way out of it. E.g. in the release notes add a note:
> The default timeout for _Run Pipeline_ has been set to 60 minutes. If that is insufficient for any of your AL-Go projects, change the value by specifying project setting `buildTimeoutInMinutes`.

[the remark this answers is not in the document]
Comment (mazhelez) [reply Actions/AL-Go-Helper.ps1:486]:
```suggestion
        "buildTimeoutInMinutes"                  = "180"
```

After getting data, it looks like 180 minutes (3 hours) is a timeout that would fit all running pipelines.

File: Scenarios/settings.md [hunk 3]
```diff
@@ -91,6 +91,7 @@ The repository settings are only read from the repository settings file (.github
 | <a id="cacheKeepDays"></a>cacheKeepDays | When using self-hosted runners, cacheKeepDays specifies the number of days docker image are cached before cleaned up when running the next pipeline.<br />Note that setting cacheKeepDays to 0 will flush the cache before every build and will cause all other running builds using agents on the same host to fail. | 3 |
 | <a id="assignPremiumPlan"></a>assignPremiumPlan | Setting assignPremiumPlan to true in your project setting file, causes the build container to be created with the AssignPremiumPlan set. This causes the auto-created user to have Premium Plan enabled. This setting is needed if your tests require premium plan enabled. | false |
 | <a id="enableTaskScheduler"></a>enableTaskScheduler | Setting enableTaskScheduler to true in your project setting file, causes the build container to be created with the Task Scheduler running. | false |
+| <a id="buildTimeoutInMinutes"></a>buildTimeoutInMinutes | An integer to specify the timeout (in minutes) for building an AL-Go project. | 60 |
```
Comment (mazhelez) [inline Scenarios/settings.md:94]:
```suggestion
| <a id="buildTimeoutInMinutes"></a>buildTimeoutInMinutes | An integer to specify the timeout (in minutes) for building an AL-Go project. | 180 |
```

**Comment by freddydk** [comment]
Do we want to rename it, since this will become the timeout for test runs in the future and not actually for the builds?
Or will we add another timeout value when we make the split?

**Review by freddydk — APPROVED** [review]

**Comment by mazhelez** [comment]
> Do we want to rename it, since this will become the timeout for test runs in the future and not actually for the builds? Or will we add another timeout value when we make the split?

It's a project setting, so I'd imagine it includes everything in the process of "building an AL-Go project". Generally, this could be a setting on the build job itself.

**Comment by freddydk** [comment]
Maybe we should add timeout on other jobs as well?
(see #521)

**Comment by mazhelez** [comment]
> Maybe we should add timeout on other jobs as well? (see #521)

That's an idea. I will look into it.

**Pull request closed without merging**